### PR TITLE
[PG Wrapper] Add diff capability

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100214

Currently we print out the mismatched collectives, but it is hard to
tell exactly the mismatch. This diff adds functionality to detect the exact mismatch
and report it.

New error is as follows:

```
Detected mismatch between collectives on ranks. Rank 0 is running collecti     ve: CollectiveFingerPrint(SequenceNumber=1151423632, OpType=ALLREDUCE, TensorShape     =[20, 10], TensorDtypes=Float, TensorDeviceTypes=TensorOptions(dtype=float (defaul     t), device=cpu, layout=Strided (default), requires_grad=false (default), pinned_me     mory=false (default), memory_format=(nullopt))), but Rank 1 is running collective:      CollectiveFingerPrint(SequenceNumber=1151423632, OpType=REDUCE, TensorShape=[20,      10], TensorDtypes=Float, TensorDeviceTypes=TensorOptions(dtype=float (default), de     vice=cpu, layout=Strided (default), requires_grad=false (default), pinned_memory=f     alse (default), memory_format=(nullopt))). Collectives differ in the following aspects:         Op type: ALLREDUCEvs REDUCE
```

i.e. the "Collectives differ in the following..." messaging is added.

Differential Revision: [D45375737](https://our.internmc.facebook.com/intern/diff/D45375737/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #include <torch/csrc/distributed/c10d/ProcessGroupWrapper.hpp>
+#include "c10d/Work.hpp"
```
Comment (H-Huang) [inline torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:2]: Is this include used? If is then it needs to have the full path to fix the build failures.

**Review by H-Huang — APPROVED** [review]
LGTM, need to fix build failures. Stamping to unblock

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp (hunk 1 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:2]: Oh weird, didn't notice because it built fine internally.

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp [hunk 2]
```diff
@@ -176,12 +177,119 @@ struct CollectiveFingerPrint {
...
+        collective_fingerprint.tensor_device_types_.size());
+    for (const auto& tensor_device_type :
+         collective_fingerprint.tensor_device_types_) {
+      device_type_strs.emplace_back(
+          c10::toString(static_cast<at::DeviceType>(tensor_device_type)));
+    }
+    return device_type_strs;
+  }
+
+  std::pair<bool, std::string> compute_collective_diff(
```
Comment (kumpera) [inline torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:228]: We could save a lot of copying by passing the stringstream as argument, but it would be a huge seekg mess in case the diff is not found.

**Review by kumpera — COMMENTED** [review]
Will this change scale when used with thousands of ranks?

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp (hunk 2 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:228]: Seems reasonable to use a stringstream, and rely on the bool return in the pair to indicate whether it is valid or not?

File: torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp [hunk 3]
```diff
@@ -176,12 +177,119 @@ struct CollectiveFingerPrint {
...
+      CollectiveFingerPrint& other) {
+    // Computes the difference between two collectives (seq num, tensor shapes,
+    // collective type, etc) for easier understanding of how mismatched
+    // collectives across ranks differ.
+    bool found_diff = false;
+    std::string ret = "Collectives differ in the following aspects: ";
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/ProcessGroupWrapper.cpp:234]: sounds good to change it to a stringstream.

**Comment by rohan-varma** [comment]
> Will this change scale when used with thousands of ranks?

I think the main issue will be that if several ranks report this mismatch issue, the logs will be very hard to parse through and read. Although the goal is that one mismatch should be able to provide enough detail about what may be going wrong with the application to help debug things.
